src/yamlprocessor/dataprocess.py

```diff
@@ -435,25 +436,43 @@ def process_variable(
             match = self.REC_SUBSTITUTE.match(tail)
             if match:
                 groups = match.groupdict()
-                substitute = groups["symbol"]
-                if len(groups["escape"]) % 2 == 0:
-                    if groups["name"] in variable_map:
-                        substitute = variable_map[groups["name"]]
-                    elif groups["name"].startswith('YP_TIME'):
+                substitute = groups['symbol']
+                if len(groups['escape']) % 2 == 0:
+                    if groups['name'] in variable_map:
+                        substitute = variable_map[groups['name']]
+                    elif groups['name'].startswith('YP_TIME'):
                         substitute = self._process_time_variable(
-                            groups["name"])
+                            groups['name'])
                     elif self.unbound_placeholder is not None:
                         substitute = str(self.unbound_placeholder)
                     else:
-                        raise UnboundVariableError(groups["name"])
-                ret += (
-                    groups["head"]
-                    + groups["escape"][0:len(groups["escape"]) // 2]
-                    + substitute)
-                tail = groups["tail"]
+                        raise UnboundVariableError(groups['name'])
+                if groups['cast']:
+                    if groups['head'] or tail != item:
+                        raise ValueError(
+                            f'{item}: bad substitution expression')
+                    try:
+                        if groups['cast'] == '.int':
+                            substitute = int(substitute)
+                        elif groups['cast'] == '.float':
+                            substitute = float(substitute)
+                        elif groups['cast'] == '.bool':
```

Comment:
Is this logic right? Won't it be inappropriately cast to True when the string is anything but 0, false, no?

Reply:
The alternate is the opposite logic. Or a strict logic of only accepting known values that can be consider boolean.

Every alternatives have some quirks.

Reply:
I agree. My preference is to accept only known values as Matt suggests.

Reply:
I am now restricting `yes`, `true` and `1` for boolean true value.